### ISSUE TITLE
Forbid deletion of required attributes

### ISSUE THREAD

**Issue by w0rp** [palantir/tslint#3146]
I have thought of something which would be a very good use of TSLint's ability to check for problems including type information, where it might not be acceptable to include a sanity check in TypeScript itself. I think it would be nice of TSLint had a rule for forbidding deletion of required attributes for objects.

Consider the following code.

```ts
const x = {foo: 3}
delete x.foo
```

The type of `x` is `{foo: number}`, but after the line which deletes the attribute `foo`, this is no longer true. TypeScript won't warn about about this at all, and allows you to delete any attribute at all.

I think TSLint should have a rule whereby deletion of optional attributes is allowed, but deletion of attributes which are non-optional is forbidden. This would help to eliminate more errors in your code.

Let me know what you think.

**Comment by ajafff** [comment]
For the reference: here's the same feature request and discussion in the TypeScript repo https://github.com/Microsoft/TypeScript/issues/13783

**Comment by JoshuaKGoldberg** [comment]
Note: per #4534, this issue will be closed in less than a month if no PR is sent to add the rule. If you _really_ need the rule, custom rules are always an option and can be maintained outside this repo!
